tiledb/sm/serialization/array_schema.cc

```diff
@@ -304,110 +306,186 @@ Status dimension_to_capnp(
+      break;
+    }
+    case Datatype::UINT64: {
+      auto val = tile_extent_reader.getUint64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::FLOAT32: {
+      auto val = tile_extent_reader.getFloat32();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::FLOAT64: {
+      auto val = tile_extent_reader.getFloat64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    default:
+      throw std::logic_error(
+          "[Deserialization::tile_extent_from_capnp] Failed to deserialize "
+          "tile extent from capnp.");
+  }
+  return tile_extent;
+}
 
+/** Deserialize a range from a cap'n proto object.
+ *
+ * @pre dim_type is valid.
+ */
+Range range_from_capnp(
+    Datatype dim_type, const capnp::Dimension::Reader& dimension_reader) {
   if (dimension_reader.hasDomain()) {
     auto domain_reader = dimension_reader.getDomain();
     Buffer domain_buffer;
-    RETURN_NOT_OK(
-        utils::copy_capnp_list(domain_reader, dim_type, &domain_buffer));
-    RETURN_NOT_OK((*dimension)->set_domain_unsafe(domain_buffer.data()));
+    Status st = utils::copy_capnp_list(domain_reader, dim_type, &domain_buffer);
+    if (!st.ok()) {
+      throw std::runtime_error(
+          "[Deserialization::range_from_capnp] Failed to deserialize domain.");
+    }
+    return Range{domain_buffer.data(), datatype_size(dim_type) * 2};
+  } else {
+    return Range{};
   }
+}
+
+/** Deserialize a dimension from a cap'n proto object. */
+shared_ptr<Dimension> dimension_from_capnp(
+    const capnp::Dimension::Reader& dimension_reader) {
+  Status st;
+
+  // Deserialize datatype
+  Datatype dim_type;
+  st = datatype_enum(dimension_reader.getType().cStr(), &dim_type);
+  if (!st.ok()) {
+    throw std::runtime_error(
+        "[Deserialization::dimension_from_capnp] " +
+        std::string(dimension_reader.getType().cStr()) +
+        " is not a valid datatype identifer.");
+  }
+
+  // Validate the dim_type, satisfying precondition of tile_extent_from_capnp
+  try {
+    ensure_datatype_is_valid(dim_type);
+  } catch (const std::exception& e) {
+    std::throw_with_nested(std::runtime_error(
+        "[Deserialization::dimension_from_capnp] " + std::string(e.what())));
+  }
+
+  // Calculate size of coordinates and re-ensure dim_type is valid (size != 0)
+  auto coord_size = datatype_size(dim_type);
+  if (coord_size == 0) {
+    throw std::logic_error(
+        "[Deserialization::dimension_from_capnp] " +
+        std::to_string(std::underlying_type_t<Datatype>(dim_type)) +
+        " is not valid as a datatype; datatype_size is 0.");
+  }
+
+  // Calculate cell_val_num
+  uint32_t cell_val_num = datatype_is_string(dim_type) ? constants::var_num : 1;
 
+  // Load domain
+  // Note: If there is no domain in capnp, a default one is constructed.
+  Range domain = Range{range_from_capnp(dim_type, dimension_reader)};
```

Comment:
Better to use braced-initialization here by eliminating the characters ` = Range`.